custom/icds/tests/test_views.py

```diff
@@ -4,12 +4,12 @@
 
 
 class TestViews(TestCase):
-    @override_settings(CUSTOM_LANDING_TEMPLATE='icds/login.html')
+    @override_settings(CUSTOM_LANDING_TEMPLATE='icds/login.html', SERVER_ENVIRONMENT='production')
```

Comment:
It seems really weird to be using `SERVER_ENVIRONMENT='production'` in custom icds tests. From the commit message:

> make tests pass even if you have overridden server_environment locally
> (if server environment is ICDS then login always redirects)

What do these tests actually test? It sounds like we have two ways to redirect login pages now? And this is testing the one that isn't even used for ICDS (and thus should probably be moved to `hqwebapp/tests`)? Does the workflow to redirect when `SERVER_ENVIRONMENT='icds'` have any tests here?

Reply:
> It seems really weird to be using `SERVER_ENVIRONMENT='production'` in custom icds tests.

Agreed. In some ways I think these tests are more about ensuring that custom ICDS workflows don't break production code (and indeed that's [why they were originally added](https://github.com/dimagi/commcare-hq/commit/d67d5fc6de19f2a1191e7289fcf8b96116eb1cf9)). No strong feelings about moving them but happy to do that in a follow up PR if you want to push for it.

In terms of the rest of your comment/question: When SERVER_ENVIRONMENT='icds' [the normal login page is short-circuited to the domain-specific login here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/views.py#L428-L430). 

That was not tested but I've added a test in 4b831d1581d7422567a2562dac60fe97abf0006e.

Reply:
(edited comment above because I was wrong about original intent of the tests)

I think these could be cleaned up to more explicitly test the various cases now that I have my head wrapped around all the branching logic that was there, though am also confident that the current tests are substantially better than what was previously there and will plan on moving on unless you want to make the case that's in important area of focus.

My main goal is to ensure the mobile login changes don't break anything moving forwards and will do my best to ensure we have test coverage that gives me (some) confidence in that.

Reply:
Thanks. I think that explanation makes sense. It can be hard to say where the integration tests should go. Keeping them in ICDS seems. It may be worth a comment on the test case about what we do intend for these to test, so that the next person who comes along understands their use